Add `QueryBuilder` support for `UNION` clause

sbuerk opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | #6368

### Summary

The `UNION` operator is used to combine the result-set
of two or more `SELECT` statements, which all database
vendors supports with usual specialities for each.

Still, there is a common shared subset which works for
all of them:

```
    SELECT column_name(s) FROM table1
    WHERE ...

    UNION <ALL | DISTINCT>

    SELECT column_name(s) FROM table2
    WHERE ...

    ORDER BY ...
    LIMIT x OFFSET y
```

with shared common requirements:

* Each `SELECT` must return the same fields
  in number, naming and order.

* Each `SELECT` **must not** have `ORDER BY`,
  expect MySQL allowing it to be used as sub
  query expression encapsulated in parenthesis.

It is now possible to build `UNION` queries using
following additional QueryBuilder API methods:

* `union(string|QueryBuilder $part)` to create a `UNION`
   query retrieving unique rows
* `addUnion(string|QueryBuilder $part, UnionType $type)`
   to add `UNION (ALL|DISTINCT)` query with the selected
   union query type.

This follows the generic logic of `select(...)` and
`addSelect(...)` along with introducing new UnionType
enum and internal QueryType::UNION enum case.

Technically, the SQL build process is dispatched to a
`DefaultUnionSQLBuilder` along with an `UnionSQLBuilder`
interface, which also allows application to implement
custom behaviour if required. Union SQL keyword and part
SQL generation is handled through added methods on the
Platforms to allow adjustment for furture versions if
needed - or throw a Exception if a Platform does not
support it anymore.

Example:

```php
$platform = $connection->getDatabasePlatform();
$qb       = $>connection->createQueryBuilder();
$select10 = $platform->getDummySelectSQL('2 as field_one');
$select20 = $platform->getDummySelectSQL('1 as field_one');
$qb->union($select10)
   ->addUnion($select20, UnionType::ALL)
   ->setMaxResults(1)
   ->setFirstResult(1)
   ->orderBy('field_one', 'ASC');
$rows = $qb->executeQuery()->fetchAllAssociative();
```

Unit and functional tests are added to demonstrate the
implementation and cover it for future changes.

Resolves: #6368

sbuerk committed:
Add `QueryBuilder` support for `UNION` clause

The `UNION` operator is used to combine the result-set
of two or more `SELECT` statements, which all database
vendors supports with usual specialities for each.

Still, there is a common shared subset which works for
all of them:

```
    SELECT column_name(s) FROM table1
    WHERE ...

    UNION <ALL | DISTINCT>

    SELECT column_name(s) FROM table2
    WHERE ...

    ORDER BY ...
    LIMIT x OFFSET y
```

with shared common requirements:

* Each `SELECT` must return the same fields
  in number, naming and order.

* Each `SELECT` **must not** have `ORDER BY`,
  expect MySQL allowing it to be used as sub
  query expression encapsulated in parenthesis.

It is now possible to build `UNION` queries using
following additional QueryBuilder API methods:

* `union(string|QueryBuilder $part)` to create a `UNION`
   query retrieving unique rows
* `addUnion(string|QueryBuilder $part, UnionType $type)`
   to add `UNION (ALL|DISTINCT)` query with the selected
   union query type.

This follows the generic logic of `select(...)` and
`addSelect(...)` along with introducing new UnionType
enum and internal QueryType::UNION enum case.

Technically, the SQL build process is dispatched to a
`DefaultUnionSQLBuilder` along with an `UnionSQLBuilder`
interface, which also allows application to implement
custom behaviour if required. Union SQL keyword and part
SQL generation is handled through added methods on the
Platforms to allow adjustment for furture versions if
needed - or throw a Exception if a Platform does not
support it anymore.

Example:

```php
$platform = $connection->getDatabasePlatform();
$qb       = $>connection->createQueryBuilder();
$select10 = $platform->getDummySelectSQL('2 as field_one');
$select20 = $platform->getDummySelectSQL('1 as field_one');
$qb->union($select10)
   ->addUnion($select20, UnionType::ALL)
   ->setMaxResults(1)
   ->setFirstResult(1)
   ->orderBy('field_one', 'ASC');
$rows = $qb->executeQuery()->fetchAllAssociative();
```

Unit and functional tests are added to demonstrate the
implementation and cover it for future changes.

Resolves: #6368

derrabus commented on docs/en/reference/query-builder.rst:
```diff
@@ -51,8 +51,8 @@ Building a Query
 
 The ``\Doctrine\DBAL\Query\QueryBuilder`` supports building ``SELECT``,
-``INSERT``, ``UPDATE`` and ``DELETE`` queries. Which sort of query you
-are building depends on the methods you are using.
+``INSERT``, ``UPDATE``, ``DELETE`` and ``UNION`` queries. Which sort of
+query you are building depends on the methods you are using.
```
There is no `UNION` query in SQL. It's technically still a `SELECT`.

sbuerk replied:
Should we avoid naming it differntly only for the documentation, or also avoid the QueryType::UNION enum case ? But than a additional flag would be required to determine that a union should be build.

derrabus replied:
_How_ we implement unions internally is an implementation detail at this point. This piece of documentation is a high level introduction that explains the purpose of the QB to someone who is familiar with SQL. Mentioning `UNION` here is simply confusing.

sbuerk replied:
Reverted that part, but added a `UNION` clause section. Hope that is better.

derrabus commented on src/SQL/Builder/DefaultUnionSQLBuilder.php:
```diff
@@ -0,0 +1,53 @@
+    public function buildSQL(UnionQuery $query): string
+    {
+        $parts      = [];
+        $modifier   = $query->isDistinct() ? ' UNION ' : ' UNION ALL ';
```
Do we really want to reuse the "distinct" flag to indicate a `UNION ALL`?

The QB's methods are usually named after the SQL syntax elements they represent in order to make them discoverable to developers. That's not really the case here.

sbuerk replied:
How would we call that than ? Or should we use dedicated methods and not having a flag method ?

* `unionAll(bool)` as flag method
* or doing it internally and have `union()` <-> `unionAll()` and `addUnion()` + `addUnionAll()` - should that be than different arrays or share one container `::$union = []`

derrabus replied:
> Or should we use dedicated methods and not having a flag method ?

Sounds good.

sbuerk replied:
Changed to `union()/addUnion()/unionAll()/addUnionAll()` way not
reusing the `distinct()` flag. Thanks for the hint, this makes
it even clearer to use it and feels better.

stof commented:
The current API with `->addUnion()` and `->addUnionAll()` is confusing IMO. The implementation will remember the DISTINCT/ALL behavior of the *last* call and apply it to all unions being added. If the SQL standard does not allow mixing both modes when having several union clauses, a separate flag would be a much better API.

I'm also wondering whether the QueryBuilder should throw exceptions when using incompatible methods together (adding select and where clauses and then using `addUnion` would silently ignore the select and where clauses right now, while the proper usage is to put those on the query builder used inside the union clause)

sbuerk commented:
> The current API with `->addUnion()` and `->addUnionAll()` is confusing IMO. The implementation will remember the DISTINCT/ALL behavior of the _last_ call and apply it to all unions being added. If the SQL standard does not allow mixing both modes when having several union clauses, a separate flag would be a much better API.
> 
> I'm also wondering whether the QueryBuilder should throw exceptions when using incompatible methods together (adding select and where clauses and then using `addUnion` would silently ignore the select and where clauses right now, while the proper usage is to put those on the query builder used inside the union clause)

@stof Thanks for your feedback.

The standard in fact allows to mix `UNION DISTINCT` and `UNION ALL` to add the next part, which was not reflected in the first API implementation.

I changed and further reduced that to `union()` and `addUnion()` methods, now dealing only with one part. The `addUnion()` enforces to define the time for the added part using a enum class. Not sure if a boolean would be more favourized over the enum, while I'm personally think the enum is more speaking in that case).

After talking with @derrabus about this PR, I moved the SQL generation into the platform classes and surround union parts on platforms supporting it (default), removing it for SQLite for now.

Platforms behave differently for additional stuff on the union select parts, which are mostly reduced using the part parenthesis surrounding excpet sqlite. Due to the nature that either string or QueryBuilder parts could be added, it is hard to scan for platform incompatible part queries - and left to be reported from the database. We have at least an excpetion in place if only one part is provided.

Not sure if we should really throw an exception if unsupported properties are set for the outer union query (for example `set()` or `values()` which is not done yet. I'd say this should be done in a dedicated change to throw exception for all types if unsupported properties and information are set beforehand and not properly cleared. What do you think ?

Not sure about the codecoverage upload error in the checks or what is needed to do to make it happy.

derrabus commented on tests/Query/QueryBuilderTest.php:
```diff
@@ -36,8 +38,18 @@ protected function setUp(): void
 
         $platform = $this->createMock(AbstractPlatform::class);
+        $platform->method('getUnionSelectPartSQL')
+            ->willReturnCallback(static function (string $sql): string {
+                return $sql;
+            });
```
```suggestion
            ->willReturnArgument(0);
```

sbuerk replied:
Ack. Done.

derrabus reviewed (commented):
LGTM, so I start nit-picking. 😉
